### PR TITLE
Matching callbacks

### PR DESCRIPTION
**Pull request by DonAurelio**
We enable a mechanism that the user can use to call callbacks that meet a given condition on an incoming upstream event (message). All the upstream events that meet the condition will call the corresponding callback. This makes easy to process several event classes with the same callback.

### CODE REVIEW

File: test/cases_sink_matching_callbacks/1.callbacks.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+typedef struct data_s data_t;
+
+void btx_initialize_usr_data(void *btx_handle, void **usr_data) {
+  data_t *data = calloc(1, sizeof(data_t));
+  *usr_data = data;
+
+  data->condition_calls_count = 0;
+  data->callback_calls_count = 0;
```
Comment (TApplencourt) [inline test/cases_sink_matching_callbacks/1.callbacks.c:17]: No need to put to 0 if calloc

File: test/test_filter_matching_callbacks.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        btx_component_label: 'D',
+        btx_component_type: 'SINK',
+        btx_component_plugin_name: 'utils',
+        btx_component_name: 'dummy',
+        # Prevent the component compilation which is not needed for babeltrace components.
+        btx_compile: false
+      }
+    ]
+
+    @btx_connect = [ 'A:C', 'B:C', 'C:D' ]
```
Comment (TApplencourt) [inline test/test_filter_matching_callbacks.rb:42]: I don't think we need to test multiple `source` for this kind of matching callbacks.
indeed this have nothing to do with the multiple source code path

**Review by TApplencourt — CHANGES_REQUESTED** [review]

File: template/upstream.c.erb [hunk 3]
```diff
@@ -68,11 +68,11 @@ void btx_matching_dispatch_<%= s.name_sanitized %>(UT_array *callbacks,
...
-                  <%= s.args.map{ |a| a.name }.join(", ") %>)) {
-      callback((void *)common_data, common_data->usr_data,
-               <%= s.args.map{ |a| a.name }.join(", ") %>);
-      *is_callback_called = true;
+
+    // is_callback_called is an output argument, indicates if the condution is met.
+    condition((void *)common_data, common_data->usr_data, is_callback_called, <%= s.args.map{ |a| a.name }.join(", ") %>);
+    if (*is_callback_called) {
+      callback((void *)common_data, common_data->usr_data,<%= s.args.map{ |a| a.name }.join(", ") %>);
     }
```
Comment (TApplencourt) [inline template/upstream.c.erb]: This is wrong.
If you register 2 callbacks,  maybe the first will pass, and the second will put it to false.

Comment (DonAurelio) [reply template/upstream.c.erb]: Agree

File: test/test_filter_matching_callbacks.rb (hunk 2 above)
Comment (DonAurelio) [reply test/test_filter_matching_callbacks.rb]: My bad, we already support that !!

File: template/upstream.h.erb [hunk 4]
```diff
@@ -17,6 +19,36 @@ typedef void
...
+typedef void
+<%= s.name_sanitized %>_conditioned_callback_f(void *btx_handle, void *usr_data,
+                                               <%= s.args.map{ |a| a.type }.join(", ") %>);
+
+extern void btx_register_matching_callback_<%= s.name_sanitized %>(void *btx_handle,
+                                            <%= s.name_sanitized %>_callback_condition_f *condition,
+                                            <%= s.name_sanitized %>_conditioned_callback_f *callback);
+<% end %>
+
+void btx_delete_matching_dispatchers(common_data_t *common_data);
```
[the remark this answers is not in the document]
Comment (DonAurelio) [reply template/upstream.h.erb:47]: I just see this comment, AFAIU extern is used for functions that will be used in other source files but their implementation will be linked after source code compilation, as is the case for `btx_register_..._callback_`.

Please correct me if I am wrong

File: template/sink.c.erb [hunk 5]
```diff
@@ -59,11 +59,27 @@ sink_consume(bt_self_component_sink *self_component_sink) {
...
+    }
+
+    {
+      const bt_stream_class* stream_class = bt_event_class_borrow_stream_class_const(event_class);
+      const char* stream_class_name = bt_stream_class_get_name(stream_class);
+
+      name_to_dispatcher_t *s = NULL;
+      HASH_FIND_STR(common_data->name_to_matching_dispatcher, stream_class_name, s);
+      if (s) {
+        (*((dispatcher_t(*))(s->dispatcher)))(s->callbacks, common_data, event, &is_callback_called);
```
Comment (TApplencourt) [inline template/sink.c.erb:80]: Let's put a comment to say that we know `is_callback_called` will only be modified if a callback is present.

File: template/upstream.h.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (TApplencourt) [reply template/upstream.h.erb:47]: Extern in a function declaration is kind of useless. It's implicitly added by default. But putting `extern` make it more obvious that it's a declaration (https://jameshfisher.com/2017/08/28/c-extern-function/)

File: test/cases_filter_matching_callbacks/1.callbacks.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+typedef struct data_s data_t;
+
+void btx_initialize_usr_data(void *btx_handle, void **usr_data) {
+  data_t *data = calloc(1, sizeof(data_t));
+  *usr_data = data;
+
+  data->condition_calls_count = 0;
+  data->callback_calls_count = 0;
```
Comment (TApplencourt) [inline test/cases_filter_matching_callbacks/1.callbacks.c:17]: No need :)

File: test/cases_filter_matching_callbacks/4.callbacks.c [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+{
+  data_t *data = (data_t *)usr_data;
+  *matched = (0 == strcmp(event_class_name, "event_1"));
+  if (*matched) data->condition_calls_count += 1;
+}
+
+static void btx_callback(void *btx_handle, void *usr_data, const char* event_class_name)
+{
+  data_t *data = (data_t *)usr_data;
+  if (0 == strcmp(event_class_name, "event_1")) data->callback_calls_count += 1;
```
Comment (TApplencourt) [inline test/cases_filter_matching_callbacks/4.callbacks.c:35]: I don't think you will need the `if` here.

Comment (TApplencourt) [reply test/cases_filter_matching_callbacks/4.callbacks.c:35]: But maybe raise an error if `event_class_name` is equal to `event_2`.

**Pull request merged**